tenant/storage_bucket.go

```diff
@@ -0,0 +1,418 @@
+	if err != nil {
+		return err
+	}
+
+	_, err = idx.Get(key)
+	// if not found then this is  _unique_.
+	if kv.IsNotFound(err) {
+		return nil
+	}
+
+	// no error means this is not unique
+	if err == nil {
+		return kv.NotUniqueError
+	}
+
+	// any other error is some sort of internal server error
+	return ErrInternalServiceError(err)
+}
+
+func unmarshalBucket(v []byte) (*influxdb.Bucket, error) {
+	u := &influxdb.Bucket{}
+	if err := json.Unmarshal(v, u); err != nil {
+		return nil, ErrCorruptBucket(err)
+	}
+
+	return u, nil
+}
+
+func marshalBucket(u *influxdb.Bucket) ([]byte, error) {
+	v, err := json.Marshal(u)
+	if err != nil {
+		return nil, ErrUnprocessableBucket(err)
+	}
+
+	return v, nil
+}
+
+func (s *Store) GetBucket(ctx context.Context, tx *Tx, id influxdb.ID) (*influxdb.Bucket, error) {
+	encodedID, err := id.Encode()
+	if err != nil {
+		return nil, InvalidOrgIDError(err)
+	}
+
+	b, err := tx.Bucket(bucketBucket)
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := b.Get(encodedID)
+	if kv.IsNotFound(err) {
+		return nil, ErrBucketNotFound
+	}
+
+	if err != nil {
+		return nil, ErrInternalServiceError(err)
+	}
+
+	return unmarshalBucket(v)
+}
+
+func (s *Store) GetBucketByName(ctx context.Context, tx *Tx, orgID influxdb.ID, n string) (*influxdb.Bucket, error) {
+	key, err := bucketIndexKey(orgID, n)
+	if err != nil {
+		return nil, &influxdb.Error{
+			Code: influxdb.EInvalid,
+			Err:  err,
+		}
+	}
+
+	idx, err := tx.Bucket(bucketIndex)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err := idx.Get(key)
+
+	// allow for hard coded bucket names that dont exist in the system
+	if kv.IsNotFound(err) {
+		switch n {
+		case influxdb.TasksSystemBucketName:
+			return &influxdb.Bucket{
+				ID:              influxdb.TasksSystemBucketID,
+				Type:            influxdb.BucketTypeSystem,
+				Name:            influxdb.TasksSystemBucketName,
+				RetentionPeriod: influxdb.TasksSystemBucketRetention,
+				Description:     "System bucket for task logs",
+				OrgID:           orgID,
+			}, nil
+		case influxdb.MonitoringSystemBucketName:
+			return &influxdb.Bucket{
+				ID:              influxdb.MonitoringSystemBucketID,
+				Type:            influxdb.BucketTypeSystem,
+				Name:            influxdb.MonitoringSystemBucketName,
+				RetentionPeriod: influxdb.MonitoringSystemBucketRetention,
+				Description:     "System bucket for monitoring logs",
+				OrgID:           orgID,
+			}, nil
+		default:
+			return nil, ErrBucketNotFound
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var id influxdb.ID
+	if err := id.Decode(buf); err != nil {
+		return nil, &influxdb.Error{
+			Err: err,
+		}
+	}
+	return s.GetBucket(ctx, tx, id)
+}
+
+type BucketFilter struct {
+	Name           *string
+	OrganizationID *influxdb.ID
+}
+
+func (s *Store) ListBuckets(ctx context.Context, tx *Tx, filter BucketFilter, opt ...influxdb.FindOptions) ([]*influxdb.Bucket, error) {
+	// this isn't a list action its a `GetBucketByName`
```

Comment:
Is this really an error condition or just a bad access pattern by the caller?

Reply:
I believe it is a pattern we would like to avoid. The reason I added the error is just so when im building the service layer I would get immediate feedback if i accidentally called list when I meant a Get then tests would fail and I could fix it quickly.